backend/canvas_app_explorer/lti1p3.py

```diff
@@ -114,14 +123,28 @@ def create_user_in_django(request: HttpRequest, message_launch: ExtendedDjangoMe
     custom_params = launch_data['https://purl.imsglobal.org/spec/lti/claim/custom']
     logger.debug(f'lti_custom_param {custom_params}')
     if not custom_params:
-        raise Exception(
-            f'You need to have custom parameters configured on your LTI Launch. Please see the LTI installation guide on the Github Wiki for more information.'
+        raise LTILaunchError(
+            'You need to have custom parameters configured on your LTI Launch. ' +
+            'Please see the LTI installation guide on the Github Wiki for more information.'
+        )
+
+    custom_param_keys = custom_params.keys()
+    if not (
+        USERNAME_KEY in custom_param_keys and
```

Comment:
I believe this [could be written](https://www.geeksforgeeks.org/python-check-if-given-multiple-keys-exist-in-a-dictionary/) as

`if not (set([USERNAME_KEY, COURSE_ID_KEY, COURSE_ROLES_KEY]).issubset(custom_params.keys())):`

You could probably also just try/except a KeyError later on 145. I prefer to write EAFP in Python rather than verifying everything. But that's an individual preference. Just trying to make have this a little less verbose.

Reply:
Yeah, I like your version better. I actually looked at `set` before cause it felt like I could do something there, but I missed this method.

I need to do some more research and soul-searching on EAFP. I think my instinct is to look before you leap (LBYL) because I find it more readable and have been burned by imprecise `try`/`except`s before, but there's probably a time and a place for both.

Reply:
https://github.com/tl-its-umich-edu/canvas-app-explorer/pull/285/commits/5f21b063c6066d60d5459afe5e495b6a9ab80005

Reply:
I read/skimmed this article just now: https://realpython.com/python-lbyl-vs-eafp/. It has some good points about using EAFP in Python, especially to minimize extra checks and prevent race condition issues.

Reply:
In Java, you basically have to NPE check almost everything but otherwise it's usually just exception handling everywhere. 

Interesting article and the point about the [zero-cost exceptions](https://github.com/python/cpython/issues/84403) is something I didn't know Python didn't have. So basically Java and Python have very inexpensive exception handling. 

I just see patterns like the one here and they seem repetitive and redundant, where they could just be solved with a try/except. This is the case for readability and repetition in the article. And there are cases to use both and not be all-in.

I think this case would fall into the uncommon and unlikely to fail side of the column. 

Use LBYL for | Use EAFP for
-- | --
Operations that are likely to fail | Operations that are unlikely to fail
Irrevocable operations, and operations that may have a side effect | Input and output (IO) operations, mainly hard drive and network operations
Common exceptional conditions that can be quickly prevented beforehand | Database operations that can be rolled back quickly

Reply:
Definitely room for me to grow here. I'll have to be more mindful of my options for exceptions in the future.